### ISSUE TITLE
HTML Head, Metatags, Seo

### ISSUE THREAD

**Issue by bnomei** [k-next/plugins#21]
for [seo](https://github.com/jenstornell/kirby-seo) @jenstornell will provide something i suppose. and maybe @pedroborges for [meta-tags v2](https://github.com/pedroborges/kirby-meta-tags).

i have a plugin port ready which provides extendable metatag support. 
https://github.com/bnomei/kirby3-htmlhead

but i am a bit afraid of creating some "field-monster" like [crafts cms seomatic](https://github.com/nystudio107/seomatic). so we should probably split it into lots of independent fields-snippets blocks? maybe we could build upon the "extendable" idea and create something together?

**Comment by jenstornell** [comment]
I will port Kirby SEO when I don't have vacation. I agree to stay away from field monsters.

**Comment by bnomei** [comment]
i am willing to help on this one, so i am assigning myself.

**Comment by pedroborges** [comment]
Meta Tags is already available for testing on a separate branch. Feedback is welcomed:

https://github.com/pedroborges/kirby-meta-tags/tree/kirby-v3

**Comment by bnomei** [comment]
@pedroborges good to hear. what about creating a "big modular plugin" together or does you plugin support most of this allready?

**Comment by bvdputte** [comment]
Imho this is a bit too much for a single plugin.

I kinda like focused plugins:
- metatags by @pedroborges : add metatags to head. These are OG, Twitter cards, Title, Description, canonical URL's and I would maybe add the favicon stuff.
- Google analytics is an entire different beast. Since this is usually added to the end of the page, is Javascript and has a whole pletora of configuration options if you consider google tag manager or measurement protocol.
- I assume Piwik / Matomo uses a different interface then GA so it would need its own plugin.
- I'm not sure what you mean with "SEO" ? If you mean the SERP-widget @jenstornell made, I'ld put this into another plugin as well, but it could use metatags as dependency?

**Comment by pedroborges** [comment]
@bvdputte Meta Tags already support favicons since they are `link` elements too, this is the config I basically use for them:

```php
c::set('meta-tags.default', function (Page $page, Site $site) {
    return [
        'link' => [
            'icon' => [
                ['href' => url('assets/images/icons/favicon-16.png'), 'sizes' => '16x16', 'type' =>'image/png'],
                ['href' => url('assets/images/icons/favicon-32.png'), 'sizes' => '32x32', 'type' =>'image/png'],
                ['href' => url('assets/images/icons/favicon-96.png'), 'sizes' => '96x96', 'type' =>'image/png'],
            ],
            'apple-touch-icon' => [
                ['href' => url('assets/images/icons/apple-touch-icon-152.png'), 'sizes' => '152x152', 'type' =>'image/png'],
                ['href' => url('assets/images/icons/apple-touch-icon-167.png'), 'sizes' => '167x167', 'type' =>'image/png'],
                ['href' => url('assets/images/icons/apple-touch-icon-180.png'), 'sizes' => '180x180', 'type' =>'image/png'],
            ],
        ],
    ];
});
```

_I have removed all other options for this example._

**Comment by jenstornell** [comment]
@bvdputte https://github.com/jenstornell/kirby-seo

Also I agree. Keep the things separated. Don't create all-in-one-solutions.

**Comment by bnomei** [comment]
@bvdputte you are right. for k2 i created my own plugin which printed everything i needed in one go extenable with snippets (not editing the config). even GA. so thats where my modular idea for "everything that goes into <head>" is based on.

what i do like about @pedroborges meta-tags plugin is that its versatile. what i do not like about it is editing everything in the config file. should i could use json_decode on json_encoded return values of snippets (you see i did that once) but it was kind of a pain.

maybe i just need a different solution on how to make the longer and longer config files easier manageable. in k2 it was possible using `include` [example](https://github.com/brocessing/kirby-webpack/blob/master/www/site/config/config.php) but with the return values structure of k3 config this is no longer possible. maybe @bastianallgeier has an idea how to split config files?

@jenstornell i agree. so this issue is solved then?

**Comment by pedroborges** [comment]
@bnomei what I do in all my projects is create a separate `site/config/meta-tags.php` file and require it in the main `site/config/config.php`. If you need to break the config into smaller parts for templates, for example, you do the same thing, even put all config in a `meta-tags` folder. You are in total control of that!

**Comment by bnomei** [comment]
@pedroborges using require will not work in k3 anymore since the config file returns an array. i would have to array_merge them manually, right?

**Comment by pedroborges** [comment]
Actually, that's possible too:

```php
<?php

return [
    'pedroborges.metatags.default' => require_once(__DIR__ . '/meta-tags/default.php'),
    'pedroborges.metatags.templates' => require_once(__DIR__ . '/meta-tags/templates.php'),
];
```

And return a closure from those files:

```php
<?php

return function ($page, $site) {
    // ...
}
```

**Comment by luxlogica** [comment]
Structured Data, IMHO, requires its own plugin. Adding structured data content for us is a requirement on every project, and the fact that there are almost 600 structure types in schema.org - most of which Google doesn't quite follow - means that we always have to spend quite a bit of time researching. 

In other CMSs, there are plugins which provide an interface in the Admin Panel, which allows the user to select from a pre-defined list of types, and then present the user with the necessary fields for them to fill out. That is exactly what we need. Here are some example use-cases - and how, in my ideal world, I'd like to define them in the page blueprint:

### Flexible Landing Page
User can generate Landing pages which can be custom-configured to be used for product promos, sales offers or special events:

``` yaml
fields:
  structured_data:
    label: Structured Data
    type: structureddata
    element:
      - product
      - offer
      - event
```

In the example above, the Panel would show a select field for "Product", "Offer" or "Event". When one of these is selected, it would then show the required fields for that data type.

### Pages with a Pre-Set, Specific Data Type
Not every page requires an option to be given to the user - some pages have a very obvious and specific data type, which then shouldn't require a selection from the user:

``` yaml
# Home Page of Local Business
fields:
  structured_data:
    label: Structured Data
    type: structureddata
    element: local-business

# Blog Article
fields:
  structured_data:
    label: Structured Data
    type: structureddata
    element: article

# Product Page in Shop
fields:
  structured_data:
    label: Structured Data
    type: structureddata
    element: product
```

In the template, the developer would be able to output the complete `<script>` element with the json-ld structured data object by using a simple method, like:

``` 
<head>
...
<?= $page->structured_data() ?>
</head>
```

**Comment by bvdputte** [comment]
@luxlogica, unless I'm misunderstanding your post, isn't what you just described just "page templates" (blueprints)? You can predefine templates your editors can choose to use (or even change into via _Settings - Change Template_).

Imho this "head metadata" plugin is something completely different.

**Comment by luxlogica** [comment]
@bvdputte Unfortunately, "page templates" are of no use in this case. 

Forgive me if my explanation was confusing - I'll try to be clearer. 'Structured Data' is markup that can be added to a page, to help search engines understand and classify the content of the page. It is used extensively by Google, and if your web page has structured data markup, Google will rank it higher. Google can also use the information in the structured data to display extra information on the page - such as images, links, a map of the business location, etc. Structured Data can be embedded into a page in a few different ways, but the recommended way is to use JSON-LD (which is just a normal ison object that follows a very specific schema). The json-ld object is placed inside a `<script>` element somewhere on the page - usually inside or close to the `<head>`. 

Depending on what the web page represents - a local business, a blog article, a product, an event, a recipe, a book, etc. - there is a specific json-ld object schema (with specific keys and structure) that must be used. All the official schemas are listed in [schema.org](https://schema.org) - and there are almost 600 of them. But out of all of those, Google - and the other major search engines - only understands and supports a handful (somewhere between 20 and 50), which are the most common.

The issue with adding structured data to a page, is that the fields you need in the structured data will vary according to the type of page/data. For a 'Local Business', the structured data has to contain the business street address, and lat/lon coordinates (so Google can try to show a map for it), logo, and so on. For a blog article, we need 'title', 'description', 'author', 'publisher' and other article info. In many cases, the information required to build the structured data can already be inferred from the content on the page. But in many cases, you cannot.

A typical example is a 'Landing Page'. We often add customisable landing pages on our sites, allowing clients to create pages as they need them. They might create landing pages that are specific to a product, or to a promotion that they're running, or to an event. They may even create landing pages that are informational articles, targeted at specific niche markets or locations. All of these landing pages can benefit greatly from having structured data markup, but the issue there is: which schema to use? If it's a landing page for a product, it should be a 'product' schema. If it's for an event, it should be an 'event' schema, and so on. And each one of these would require the client to fill out a different number of fields.

If we're using _blueprints_ to do this, we'd have to create one blueprint for each type of possible schema to be used. That's not what is wanted. What we need in this scenario is to have a dropdown menu (select field) in the panel, where the client can choose themselves which schema they need for that page: 'none', 'article', 'local business', 'product', 'event', etc. Once they select the schema, then the appropriate fields are displayed for them to fill out.

**Comment by luxlogica** [comment]
@bvdputte indeed, what we need in order for a 'structured data' plugin like this to exist is a 'conditional field'. Outputting the data to the page is easy! ;-)
